src/screens/Distribution/RingChart.js

```diff
@@ -0,0 +1,143 @@
+// @flow
+
+import React, { PureComponent } from "react";
+import { PanResponder, View } from "react-native";
+import * as d3shape from "d3-shape";
+import Svg, { Path, G, Circle } from "react-native-svg";
+import colors from "../../colors";
+import type { DistributionItem } from "./DistributionCard";
+
+type Props = {
+  data: Array<DistributionItem>,
+  side: number,
+  style?: *,
+  selectedKey: number => void,
+  highlight?: number,
+};
+
+class RingChart extends PureComponent<Props> {
+  static defaultProps = {
+    data: [],
+  };
+
+  panResponder: *;
+  arcGenerator = d3shape.arc();
+  offsetX = 0;
+  offsetY = 0;
+  highlight = -1;
+  paths: * = {};
+
+  constructor(props: Props) {
+    super(props);
+    const { highlight, side, selectedKey, data } = this.props;
+
+    if (highlight) {
+      this.highlight = highlight;
+    }
+
+    this.panResponder = PanResponder.create({
+      onPanResponderGrant: event => {
+        this.offsetX = event.nativeEvent.pageX - event.nativeEvent.locationX;
+        this.offsetY = event.nativeEvent.pageY - event.nativeEvent.locationY;
+      },
+      onStartShouldSetPanResponder: () => true,
+      onStartShouldSetPanResponderCapture: () => false,
+      onMoveShouldSetPanResponder: () => false,
+      onMoveShouldSetPanResponderCapture: () => false,
+
+      onPanResponderMove: event => {
+        const x = event.nativeEvent.pageX - this.offsetX;
+        const y = event.nativeEvent.pageY - this.offsetY;
+
+        const y0 = side / 2;
+        const x0 = side / 2;
+        let rad = Math.atan2(y - y0, x - x0);
+        rad =
+          ((rad > 0 ? rad : 2 * Math.PI + rad) + 1.5707963268) % (2 * Math.PI);
+
+        let highlight = this.highlight;
+        for (let i = 0; i < this.paths.items.length; i++) {
+          if (this.paths.items[i].endAngle > rad) {
+            highlight = i;
+            break;
+          }
+        }
+        if (this.highlight !== highlight) {
+          selectedKey(highlight);
+        }
+      },
+
+      onShouldBlockNativeResponder: () => false,
+    });
+    this.paths = data.reduce(this.reducer, {
```

Comment:
this shouldn't be done in the constructor.

strictly speaking, it is a derivate data of the data props so it normally should live in `render()` method itself (what is data changes).

As you need this also to be used in constructor, what you can do is put that `this.paths=` logic in a method, and call if from constructor AND in `componentDidUpdate`